### DISCUSSION TITLE
esp8266 client port

### DISCUSSION THREAD

**Discussion by 532910** [snapcast/snapcast#347] [General]
It will be very cool to make clients on esp8266.
This is feature request of course

**Comment by isaactzab** [comment]
I think also that if works at esp8266 will be amazing but esp8266 specs are poor to handle all project. Please correct me if I am wrong.

**Comment by isaactzab** [comment]
Maybe don't esp8266, but another cheap but little more powerful chipset. Some idea?

**Comment by localhoster** [comment]
You need a device which can handle a full OS to run Snapcast, as it has alot of dependencies. ESP8266 is a microcontroller and has nothing to do Snapcast. What you definetely can use are  Raspberrie/Orange/Banana Pis in different versions. Some of them, like OPi Zero are very cheap and work well with Snapcast.

**Comment by isaactzab** [comment]
Where to buy a cheap orange pi? Aliexpress price is $7 (+$5 of shipping 40-50 days to my country)

**Comment by palitu** [comment]
as far as i have found, the OPi Zero is the cheapest SBC with a case that
also has native 3.5mm output.

Otherwise, i go for a RPi Zero W, with an Suptronix X400 amplifier for ~
AUD $50 solution, but it doesnt have a case.

**Comment by 532910** [comment]
but OPi Zero doesn't have i2s
What is the cheapest board with i2s?

**Comment by palitu** [comment]
OPi has direct output for an audio jack.

If you need i2s, i have used an RPi Zero W.  There are probably others
though.

**Comment by jaecksch** [comment]
esp8266 is not powerfull enough. An alternative would be the successor ESP32. But on ESP32 there is only RTOS as operating system available and all the stuff and libs need to be ported.  I tried some time ago very lightweight Openwrt boards like RT3550 (olimex) and Linkit Smart 7866 (mediatek). Both have wireless on board and have an I2S audio interface. Snapclient worked on both, but only with USB speakers. On RT3550 I was not able to get the I2S driver to run and on Linkit board I got stuttering sound. Now I step back to raspberry pi which works very well.

**Comment by 532910** [comment]
Will Onion Omega2 be enough?

**Comment by isaactzab** [comment]
Looks good. 
> Will Onion Omega2 be enough?

**Comment by palitu** [comment]
The problem that i have found with the onion is that it gets expensive for
the other hardware around it.  the microcontroller part is cheap, but to do
anything else it starts to add up.

**Comment by isaactzab** [comment]
maybe chip computer.
https://getchip.com/pages/chip

**Comment by isaactzab** [comment]
oh sorry; $9 chip is not available.

> maybe chip computer.
https://getchip.com/pages/chip

**Comment by XDjackieXD** [comment]
I'm successfully running snapclient on an [A5-11](https://wiki.openwrt.org/toh/unbranded/a5-v11?s[]=rt5350f) router with a custom, stripped-down LEDE build (no LUCI, no Wifi drivers, ... as the flash is really small) and a 3€ usb soundcard (one that only does digital out as the sound system in my living room supports it). that way the complete receiver costs ~7€ which is quite a good price point.

**Reply by sevba** [reply]
Could you tell us how you did the LEDE build?

**Reply by XDjackieXD** [reply]
https://openwrt.org/docs/guide-developer/toolchain/use-buildsystem like any other openwrt built but with some additions that are usually not part of a default openwrt build (usb audio, alsa, alsamixer and snapcast)

**Reply by StuartIanNaylor** [reply]
Do you have a repo?

**Reply by XDjackieXD** [reply]
No I can't find my old build config either but if your device is supported by openwrt it's straight forward (setup the build env for your device as described in the linked wiki page, run `make menuconfig`, select alsa, usb audio drivers and snapcast and build it. to save on space for devices with small flash, you can remove all the luci components and uhttpd)

**Reply by StuartIanNaylor** [reply]
I think I might search one out on aliexpress and give it a try

**Reply by XDjackieXD** [reply]
I remember this device only having 32MB of RAM. I think recent OpenWRT versions wont be stable with only so little RAM.
There are some other cheap and compact, way newer devices with 64MB of RAM that will probably work better.

**Reply by XDjackieXD** [reply]
Sadly the NEXX WT3020F is not available anymore (it was a ~10€ OpenWRT mini router) and the GL-MT300N-V2 went up in price significantly in the last year :(
To be a bit more specific about the 32MB RAM thing: OpenWRT does not support 4MB Flash / 32MB RAM devices officially since 2019 anymore. You *can* get recent OpenWRT to work on them but it requires some work and might not be stable (removing all the non-needed kernel compile options, enabling compressed RAM, removing many non-needed OpenWRT components...).

**Reply by sevba** [reply]
Well it's going to be running on my isolated Network-of-Things VLAN, purely for Snapclient. For that use case I'll accept the risk of using a pre-2019 OpenWRT firmware. :)
It's [this](https://www.aliexpress.com/item/32842478867.html) one, right?

**Reply by StuartIanNaylor** [reply]
Pine supposedly are doing something that blurs the microcontroller / soc divide with https://www.cnx-software.com/2022/10/10/pine64-ox64-sbc-bl808-risc-v-multi-protocol-wisoc-64mb-ram/ 
I wonder if those would be any good for a Openwrt Snapclient?

**Reply by MM2forever** [reply]
I know this is a very old thread but I am biting my teeth out for over a week now trying to compile an openwrt image for the A5-V11 and I have tried so many different ways and they all fail at one thing or another.
Is there any chance one of you has a working sysupgrade image with snapcast client and USB audio laying around to share or can pull it from an existing device?
I don't want to be defeated :(

**Comment by ghost** [comment]
I have some ESP8266 and ESP32 I use in my smarthome project: they are both definitely powerful enough for the task of running a snapcast client, but it would have to be a rewrite, not a port, running on RTOS or built in arduino IDE.
What's good with these is the native i2s interface, which means it's way easier to get good audio than on the raspberry, with sound boards available for about 3€.
The main limit of the ESP8266 for this task is bandwidth IMHO: hardly 10Mbps in ideal conditions, far less if it's busy doing else... ESP32 is way better, with a dual core CPU and 135Mbps datasheet speed, benchmarking in excess of 40Mbps.

A word on fancy SBC: they all look great on their websites, but many of them are almost impossible to find, and if you find them the price is at least twice what promised. After that, you'll definiitely still have quality and design issues, since something as well supported and expensive as the Raspberry Pi 3 still can't manage to get wifi and bluetooth working well simoultaneusly, or a reliable Wifi on its own :-(

My 2Cents...

**Comment by isaactzab** [comment]
So this could be the begin of new project to look for very cheap multiroom audio based on snapcast?

**Comment by frostworx** [comment]
In most rooms I use TP-Link TL-WR710N wallplugs with Logitech S150 soundboxes as audio devices here.
OS is openwrt/lede and the logitech boxes are plugged into the tplink with just one usb cable (so audio and power are provided by one cable). you can also use triggerhappy on lede/openwrt to use the buttons on one of the boxes to control the audio (volume up/down and dis-/connect from snapcast server using netcat with the mute button).

**Comment by araczkowski** [comment]
@frostworx very interesting and creative approach :+1: 
But how you did this? could you please share more details?

I'm trying to do the same on Gargoyle 1.10.0.3 from https://dl.eko.one.pl/firmware/?vendor=TP-Link&model=TL-WR710N&revision=v1&imageType=factory&showFirmwareTable=false

and snapclient_0.12.0_ar71xx.ipk from https://github.com/badaix/snapcast/releases
but the answer is always: "Not enough space in destination root to install specified packages"

full log:
```
root@Gargoyle:~# wget https://github.com/badaix/snapcast/releases/download/v0.12.0/snapclient_0.12.0_ar71xx.ipk
Connecting to github.com (192.30.253.113:443)
Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (54.231.120.187:443)
snapclient_0.12.0_ar 100% |******************************************************************************************************|   144k  0:00:00 ETA

root@Gargoyle:~# opkg update
Downloading package list for chaos_calmer_base source...
Package list for chaos_calmer_base downloaded successfully.

Downloading package list for chaos_calmer_packages source...
Package list for chaos_calmer_packages downloaded successfully.

Downloading package list for chaos_calmer_telephony source...
Package list for chaos_calmer_telephony downloaded successfully.

Downloading package list for gargoyle_pl source...
Package list for gargoyle_pl downloaded successfully.

Downloading package list for chaos_calmer_routing source...
Package list for chaos_calmer_routing downloaded successfully.

Downloading package list for eko_one_pl source...
Package list for eko_one_pl downloaded successfully.

Downloading package list for chaos_calmer_management source...
Package list for chaos_calmer_management downloaded successfully.

root@Gargoyle:~# opkg install snapclient_0.12.0_ar71xx.ipk 
ERROR: Not enough space in destination root to install specified packages:
	libstdcpp, libflac, kmod-input-core, alsa-lib, libogg, dbus, snapclient, libavahi-dbus-support, libatomic, libavahi-client, libvorbisidec, libdbus, libdaemon, kmod-sound-core, avahi-dbus-daemon, libexpat
```

of-course probably I need to have more space :) but I don't know what can I uninstall/remove 

any advice?

**Comment by frostworx** [comment]
@araczkowski,
currently my players have a selfrolled lede musl snapshot installed. afair I had a (also self rolled) openwrt designated driver before also using musl. iirc I also had problems on chaos calmer and afair the snapclient ipk is using uclibc. never used gargoyle so no idea if the ipk is compatible. maybe you could scp the snapclient binary to /tmp/ and test if it even would work. good luck so far!

**Comment by frostworx** [comment]
some might be intererested in:
https://github.com/espressif/esp-adf

**Comment by ash-vd** [comment]
I use a RPI Zero W with a JustBoom AMP zero for most rooms. This combination is small enough for me to build into the speakers, so the only connection to the speaker is a power cord. I am still looking for a cheap 5ghz wifi usb dongle which works on Raspbian. For now I use the onboard ones, but the reception could be better.

**Comment by palitu** [comment]
> 
> 
> I use a RPI Zero W with a JustBoom AMP zero for most rooms. This combination is small enough for me to build into the speakers, so the only connection to the speaker is a power cord. I am still looking for a cheap 5ghz wifi usb dongle which works on Raspbian. For now I use the onboard ones, but the reception could be better.

Hey @ashwinvandijk, do you have any special build instructions?  or cases?

I have done that too, but i would prefer a more out of the box solution, where i burn the image and when it starts i can connect to it via its own AP, setup WiFi details and what not, then it "just works" sort of thing.

I think that this is the most affordable way to get a new node built.

**Comment by StuartIanNaylor** [comment]
> 5ghz wifi usb dongle

https://www.aliexpress.com/item/4000079170817.html ?

I have been using the $10 WM8960 soundcards as widearray network mics are interesting but the Respeaker/Waveshare modules are pretty good with TPA3116 amps that give 40watt+ from 24V.
Just a shame they are fed from the headphone amp rather than line out but pretty good.

The RK3308 boards from Radxa are as cheap even cheaper than ESP and have a codec built in, but run debian.

**Comment by MayeulC** [comment]
Some might be interested in this repository I found while looking for the same thing: https://github.com/jorgenkraghjakobsen/snapclient

There are a couple of interesting approaches in this thread, they are worth investigating. But having it integrated with esphome/homeassistant would be a huge plus!

**Comment by jblb** [comment]
Hi,
I am not an expert but  https://github.com/earlephilhower/ESP8266Audio could help

**Comment by zalerion** [comment]
> 
> I have done that too, but i would prefer a more out of the box solution, where i burn the image and when it starts i can connect to it via its own AP, setup WiFi details and what not, then it "just works" sort of thing.
> 
> I think that this is the most affordable way to get a new node built.

Anything like opening an own AP and setup stuff will slow down the device! I have tried something similar, not for setup but for automatic switching if no WiFi is available. Basically: Not at home -> stop client, start AP, start local MPD, enable BT for streaming from phone... Basically you could setup a server this way for mobile speakers, where others connect to the main one for syncing :D

Never used it so now it is just a plain simple snapclient. But it is possible.

On the other hand: Anything you would do this way, would have to be an own image for flashing. As long as you use the same hardware every time (and didn't make it read only), just pull the image from a working node. SSH to it to change its name. Done.

**Comment by douardda** [comment]
Just in case it helps: https://github.com/jorgenkraghjakobsen/snapclient

**Comment by StuartIanNaylor** [comment]
Prob would be much easier for https://github.com/jorgenkraghjakobsen/snapclient is the 8mb esp32-s3 was used as also now starting to get super cheap but the vector instructions give a big boost to FFT and all the codecs it could run and the mem space will prob help much with the sync.

Would be so great if Jorgen and others picked up on this there is the new esp32-p4 but that will be some time and not sure if it will be much faster than the s3 but just without cadence IP.

I am going to give a pre-2019 OpenWRT firmware a try with a https://openwrt.org/toh/unbranded/a5-v11 even if outdated.

**Reply by MayeulC** [reply]
I've been thinking of toying with an [Ox64](https://pine64.com/product-category/ox64/) lately. It has an I2S interface, can run Linux if necessary, with 64MB of RAM and a few RISC-V cores (64 and 32 bit).

**Reply by StuartIanNaylor** [reply]
Yeah they looked really good but the shipping for me took it to $30 and that sort of stopped me.

**Reply by douardda** [reply]
But I don't see anything related to I2S in the [BL808 MCU SDK](https://github.com/bouffalolab/bl_mcu_sdk) so no idea how feasible it would be... Seems a bit too new to really be usable for now.

**Reply by StuartIanNaylor** [reply]
Plus anyway for the cost + delivery for a [Ox64](https://pine64.com/product-category/ox64/) I can get a OrangePi02 for the same $30

I used to call OrangePi & BananPi the 'strange fruit' but after purchasing a OrangePi5 I have revised my expectations that where very low.
I have a OrangePi02 still to be delivered so will give you guys a review with the current Pi situation, still would be great to get a ultra low cost / low energy and think likely the esp32-s3 16mb/8mb would likely fit where I think the standard ESP32 got a bit cramped.
I just wish they did a dev kit that was simple but just with ADC/DAC maybe the ones off the esp32-s3-box.
